File: core/managers/ai/tool-dispatcher.ts
/**
 * ToolDispatcher — 도구 호출 사전 검증 + dispatch.
 *
 * AiManager 의 내부 collaborator (외부 import 금지).
 *
 * 책임:
 *   1. resolveCallTarget — AI 가 호출한 identifier (변형 포함) → 실제 dispatch target.
 *      MCP 서버명·system/user 모듈 경로 매칭. Core.resolveCallTarget 도 이 메서드 위임.
 *   2. executeToolCall — ToolManager 등록 도구 → render_* 변형 정규화 → resolveCallTarget 분기.
 *   3. checkNeedsApproval — schedule_task / write_file / save_page 등 사전 승인 필요 도구 식별.
 *   4. preValidatePendingArgs — 승인 대기 도구 인자 검증 (필수 필드 누락·잘못된 평면 인자 등).
 *
 * 분리 이유: dispatch 결정이 멀티턴 루프 본체와 독립. Core facade 가 직접 호출하는 resolveCallTarget 도 포함.
 *
 * 일반 로직: 도구별 enumerate 가 아닌 패턴 매칭 + 자동 정규화.
 */
import type { FirebatCore } from '../../index';
import type { AiRequestOpts } from '../../index';
import type { ToolCall } from '../../ports';
import { RENDER_TOOL_MAP, normalizeRenderName } from '../../../lib/render-map';

const CALL_TARGET_TTL = 60_000;

type CallTarget = { kind: 'mcp'; server: string } | { kind: 'execute'; path: string };

export class ToolDispatcher {
  private callTargetCache: { map: Map<string, CallTarget>; ts: number } | null = null;

  constructor(private readonly core: FirebatCore) {}

  /** Core.resolveCallTarget 이 위임 — AI 가 다양한 변형으로 호출해도 자동 매칭.
   *  매칭 우선순위: 정확한 이름 → snake/kebab 변형 → sysmod_ 접두사 / full path → null.
   *  60초 캐시 — listDir / listMcpServers 호출 비용 절감. */
  async resolveCallTarget(identifier: string): Promise<CallTarget | null> {
    if (!identifier) return null;
    const lookup = (id: string, map: Map<string, CallTarget>) =>
      map.get(id) ?? map.get(id.replace(/_/g, '-')) ?? map.get(id.replace(/-/g, '_'));
    if (this.callTargetCache && (Date.now() - this.callTargetCache.ts) < CALL_TARGET_TTL) {
      const hit = lookup(identifier, this.callTargetCache.map);
      if (hit !== undefined) return hit;
    }
    const map = new Map<string, CallTarget>();
    // 1) 외부 MCP 서버
    try {
      const mcpServers = this.core.listMcpServers();
      if (Array.isArray(mcpServers)) {
        for (const s of mcpServers) {
          if (!s?.name) continue;
          const target: CallTarget = { kind: 'mcp', server: s.name };
          map.set(s.name, target);
          map.set(s.name.replace(/-/g, '_'), target);
          map.set(s.name.replace(/_/g, '-'), target);
        }
      }
    } catch { /* MCP 미설정 무시 */ }
    // 2) system + user modules
    for (const dir of ['system/modules', 'user/modules']) {
      try {
        const ls = await this.core.listDir(dir);
        if (!ls.success || !ls.data) continue;
        for (const e of ls.data.filter(x => x.isDirectory)) {
          const path = `${dir}/${e.name}/index.mjs`;
          const target: CallTarget = { kind: 'execute', path };
          map.set(e.name, target);
          map.set(e.name.replace(/-/g, '_'), target);
          map.set(e.name.replace(/_/g, '-'), target);
          map.set(`sysmod_${e.name}`, target);
          map.set(`sysmod_${e.name.replace(/-/g, '_')}`, target);
          map.set(path, target);
        }
      } catch { /* 폴더 없음 무시 */ }
    }
    this.callTargetCache = { map, ts: Date.now() };
    return lookup(identifier, map) ?? null;
  }

  /** 단일 도구 호출 실행 — 결과를 Record<string, unknown>로 반환.
   *  순서: ToolManager 등록 도구 → render_* 변형 정규화 → resolveCallTarget (sysmod/mcp 자동 분기) → 알 수 없는 도구 에러. */
  async executeToolCall(tc: ToolCall, opts?: AiRequestOpts): Promise<Record<string, unknown>> {
    try {
      // 1) ToolManager 등록 도구 — 정적·동적 모두 단일 dispatch.
      if (this.core.getToolDefinition(tc.name)) {
        return await this.core.executeTool(tc.name, tc.args as Record<string, unknown>, {
          conversationId: opts?.conversationId,
          owner: opts?.owner,
          requestOpts: opts as Record<string, unknown> | undefined,
        });
      }
      // 2) render_* 변형 정규화 — AI 가 'table' / 'render-chart' 등으로 불러도 매칭.
      const renderName = normalizeRenderName(tc.name);
      if (renderName && RENDER_TOOL_MAP[renderName]) {
        return { success: true, component: RENDER_TOOL_MAP[renderName], props: tc.args as Record<string, unknown> };
      }
      // 3) 통합 resolver — sysmod / mcp 자동 분기.
      const target = await this.resolveCallTarget(tc.name);
      if (target?.kind === 'execute') {
        const res = await this.core.sandboxExecute(target.path, tc.args);
        if (!res.success) return { success: false, error: res.error };
        if (res.data?.success === false) return { success: false, error: JSON.stringify(res.data) };
        return { success: true, data: res.data };
      }
      // mcp_{server}_{tool} 접두사 — server/tool 분리.
      if (tc.name.startsWith('mcp_')) {
        const parts = tc.name.slice(4).split('_');
        const server = parts[0];
        const tool = parts.slice(1).join('_');
        const res = await this.core.callMcpTool(server, tool, tc.args);
        return res.success ? { success: true, data: res.data } : { success: false, error: res.error };
      }
      if (target?.kind === 'mcp') {
        return { success: false, error: `MCP 서버 '${target.server}' 호출 시 도구 이름 명시 필요 (예: mcp_${target.server}_{tool} 형태).` };
      }
      return { success: false, error: `알 수 없는 도구: ${tc.name}` };
    } catch (e: any) {
      return { success: false, error: e.message };
    }
  }

  /** 사전 승인 필요 여부 판정 — 되돌리기 어려운 작업만 user confirmation. 일반 로직, 도구별 분기 X.
   *  null 반환 = 즉시 실행 OK. { summary } = pending action 으로 UI 표시. */
  async checkNeedsApproval(tc: ToolCall): Promise<{ summary: string } | null> {
    switch (tc.name) {
      case 'write_file': {
        const path = (tc.args as { path?: string }).path;
        if (!path) return null;
        const exists = await this.core.readFile(path);
        if (!exists.success) return null; // 새 파일은 즉시 작성
        return { summary: `파일 수정: ${path}` };
      }
      case 'save_page': {
        const slug = (tc.args as { slug?: string }).slug;
        if (!slug) return null;
        const exists = await this.core.getPage(slug);
        if (!exists.success) return null; // 새 페이지는 즉시 저장
        return { summary: `페이지 수정: /${slug}` };
      }
      case 'delete_file': {
        const path = (tc.args as { path?: string }).path;
        return { summary: `파일 삭제: ${path ?? '(unknown)'}` };
      }
      case 'delete_page': {
        const slug = (tc.args as { slug?: string }).slug;
        return { summary: `페이지 삭제: /${slug ?? '(unknown)'}` };
      }
      case 'schedule_task': {
        const args = tc.args as { title?: string; cronTime?: string; runAt?: string; delaySec?: number };
        const when = args.cronTime ?? args.runAt ?? (args.delaySec != null ? `${args.delaySec}초 후` : '');
        return { summary: `예약 등록: ${args.title ?? '(제목 없음)'} (${when})` };
      }
      default:
        return null;
    }
  }

  /** 승인 대기 도구 인자 사전 검증 — 실패 시 에러 메시지 반환 (pending 생성 전 거부).
   *  잘못된 인자로 pending 만드는 걸 차단 — UI 가 무조건 "승인" 버튼 보여주는 헛발질 방지. */
  preValidatePendingArgs(tc: ToolCall): string | null {
    const args = tc.args as Record<string, unknown>;
    switch (tc.name) {
      case 'schedule_task': {
        const isAgent = args.executionMode === 'agent';
        const hasTarget = typeof args.targetPath === 'string' && (args.targetPath as string).trim() !== '';
        const hasPipeline = Array.isArray(args.pipeline) && (args.pipeline as unknown[]).length > 0;
        const hasAgentPrompt = typeof args.agentPrompt === 'string' && (args.agentPrompt as string).trim() !== '';
        // agent 모드: agentPrompt 필수. pipeline 모드: targetPath/pipeline 중 하나 필수.
        if (isAgent) {
          if (!hasAgentPrompt) {
            return 'schedule_task 인자 누락: agent 모드는 agentPrompt 필수입니다. 트리거 시 AI 에 전달할 자연어 instruction (잡 목적·필요 데이터·출력 형식·알림) 작성하세요.';
          }
        } else if (!hasTarget && !hasPipeline) {
          return 'schedule_task 인자 누락: targetPath 또는 pipeline 중 하나는 반드시 지정해야 합니다. (agent 모드면 executionMode:"agent" + agentPrompt)';
        }
        const hasWhen = !!args.cronTime || !!args.runAt || args.delaySec != null;
        if (!hasWhen) return 'schedule_task 인자 누락: cronTime / runAt / delaySec 중 하나는 반드시 지정해야 합니다.';
        if (hasPipeline) {
          const pipeline = args.pipeline as unknown[];
          for (let i = 0; i < pipeline.length; i++) {
            const step = pipeline[i] as Record<string, unknown> | null;
            if (!step || typeof step !== 'object') return `[Step ${i + 1}] step이 객체가 아닙니다.`;
            const t = step.type;
            if (!t || typeof t !== 'string') return `[Step ${i + 1}] type 누락 — EXECUTE/MCP_CALL/NETWORK_REQUEST/LLM_TRANSFORM/CONDITION 중 하나를 지정하세요.`;
            if (!['EXECUTE', 'MCP_CALL', 'NETWORK_REQUEST', 'LLM_TRANSFORM', 'CONDITION'].includes(t)) {
              return `[Step ${i + 1}] 알 수 없는 type: ${t}`;
            }
            if (t === 'EXECUTE') {
              if (!step.path) return `[Step ${i + 1}] EXECUTE에 path 필수 (예: system/modules/kakao-talk/index.mjs).`;
              const id = step.inputData as Record<string, unknown> | undefined;
              if (!id || typeof id !== 'object' || Object.keys(id).length === 0) {
                return `[Step ${i + 1}] EXECUTE 인자 오류: 모듈 실행 파라미터는 step 평면이 아니라 inputData 객체에 넣어야 합니다. 잘못: {type:"EXECUTE",path:"...",action:"price",symbol:"..."} · 올바름: {type:"EXECUTE",path:"...",inputData:{action:"price",symbol:"..."}}`;
              }
            }
            if (t === 'MCP_CALL' && (!step.server || !step.tool)) return `[Step ${i + 1}] MCP_CALL에 server, tool 필수.`;
            if (t === 'NETWORK_REQUEST' && !step.url) return `[Step ${i + 1}] NETWORK_REQUEST에 url 필수.`;
            if (t === 'LLM_TRANSFORM' && !step.instruction) return `[Step ${i + 1}] LLM_TRANSFORM에 instruction 필수.`;
            if (t === 'CONDITION' && (!step.field || !step.op)) return `[Step ${i + 1}] CONDITION에 field, op 필수.`;
          }
        }
        return null;
      }
      case 'write_file': {
        if (typeof args.path !== 'string' || !(args.path as string).trim()) return 'write_file 인자 누락: path 필수.';
        if (args.content == null) return 'write_file 인자 누락: content 필수.';
        return null;
      }
      case 'save_page': {
        if (typeof args.slug !== 'string' || !(args.slug as string).trim()) return 'save_page 인자 누락: slug 필수.';
        if (!args.spec || typeof args.spec !== 'object') return 'save_page 인자 누락: spec 필수.';
        return null;
      }
      default:
        return null;
    }
  }
}

File: core/src/managers/ai/tool_dispatcher.rs
//! ToolDispatcher — 도구 호출 사전 검증 + 승인 게이트.
//!
//! 옛 TS `core/managers/ai/tool-dispatcher.ts` 1:1 port.
//!
//! 책임:
//!   1. `resolve_call_target(identifier)` — AI 가 호출한 변형 (snake/kebab/sysmod_) → 실제 dispatch target.
//!      MCP 서버명·system/user 모듈 경로 매칭. 60초 캐시 (listMcpServers / list_dir 호출 비용 절감).
//!   2. `check_needs_approval(tool_call)` — write_file / save_page / delete_* / schedule_task /
//!      cancel_task 6 케이스. 되돌리기 어려운 작업만 user confirmation.
//!   3. `pre_validate_pending_args(tool_call)` — schedule_task / write_file / save_page 사전 검증.
//!      잘못된 인자로 pending 만드는 헛발질 차단.
//!
//! 분리 이유: dispatch 결정이 멀티턴 루프 본체와 독립.
//! 일반 로직: 도구별 enumerate 가 아닌 패턴 매칭 + 자동 정규화.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use crate::managers::mcp::McpManager;
use crate::managers::page::PageManager;
use crate::managers::schedule::ScheduleManager;
use crate::ports::{IStoragePort, ToolCall};

const CALL_TARGET_TTL: Duration = Duration::from_secs(60);

/// 도구 dispatch 의 실제 대상 — MCP 서버 또는 sandbox execute path.
#[derive(Debug, Clone)]
pub enum CallTarget {
    /// 외부 MCP 서버 — `mcp_<server>_<tool>` 패턴으로 호출.
    Mcp { server: String },
    /// system/user 모듈 — sandbox.execute(path) 으로 호출.
    Execute { path: String },
}

/// 사전 승인 필요 도구의 사용자 표시 요약.
#[derive(Debug, Clone)]
pub struct ApprovalSummary {
    pub summary: String,
}

struct CachedTargets {
    map: HashMap<String, CallTarget>,
    cached_at: Instant,
}

pub struct ToolDispatcher {
    storage: Arc<dyn IStoragePort>,
    /// 옵션 — 박혀있으면 cancel_task / save_page / 외부 MCP 검증 활성. 미박음 시 검증 skip.
    page: Option<Arc<PageManager>>,
    schedule: Option<Arc<ScheduleManager>>,
    mcp: Option<Arc<McpManager>>,
    /// 60초 캐시 — listMcpServers / list_dir 비용 절감.
    cache: Mutex<Option<CachedTargets>>,
}

impl ToolDispatcher {
    pub fn new(storage: Arc<dyn IStoragePort>) -> Self {
        Self {
            storage,
            page: None,
            schedule: None,
            mcp: None,
            cache: Mutex::new(None),
        }
    }

    pub fn with_page(mut self, page: Arc<PageManager>) -> Self {
        self.page = Some(page);
        self
    }

    pub fn with_schedule(mut self, schedule: Arc<ScheduleManager>) -> Self {
        self.schedule = Some(schedule);
        self
    }

    pub fn with_mcp(mut self, mcp: Arc<McpManager>) -> Self {
        self.mcp = Some(mcp);
        self
    }

    /// 변형 매칭 헬퍼 — `id` / `id.replace(_, -)` / `id.replace(-, _)` 셋 다 시도.
    fn lookup_variants(map: &HashMap<String, CallTarget>, id: &str) -> Option<CallTarget> {
        if let Some(t) = map.get(id) {
            return Some(t.clone());
        }
        let snake = id.replace('-', "_");
        if let Some(t) = map.get(&snake) {
            return Some(t.clone());
        }
        let kebab = id.replace('_', "-");
        if let Some(t) = map.get(&kebab) {
            return Some(t.clone());
        }
        None
    }

    /// `resolve_call_target` — AI 가 다양한 변형으로 호출해도 자동 매칭.
    /// 매칭 우선순위: 정확한 이름 → snake/kebab 변형 → sysmod_ 접두사 / full path → None.
    /// 60초 캐시 — listMcpServers / list_dir 호출 비용 절감.
    pub async fn resolve_call_target(&self, identifier: &str) -> Option<CallTarget> {
        if identifier.is_empty() {
            return None;
        }

        // 캐시 hit 검사
        {
            let cache = self.cache.lock().ok()?;
            if let Some(cached) = cache.as_ref() {
                if cached.cached_at.elapsed() < CALL_TARGET_TTL {
                    if let Some(t) = Self::lookup_variants(&cached.map, identifier) {
                        return Some(t);
                    }
                    // cache 안에 미존재 — 새로 빌드 시도 (cache 만료 체크는 위에서)
                    return None;
                }
            }
        }

        // 새 캐시 빌드
        let mut map: HashMap<String, CallTarget> = HashMap::new();

        // 1) 외부 MCP 서버
        if let Some(mcp) = &self.mcp {
            for server in mcp.list_servers() {
                let name = server.name.clone();
                if name.is_empty() {
                    continue;
                }
                let target = CallTarget::Mcp {
                    server: name.clone(),
                };
                map.insert(name.clone(), target.clone());
                map.insert(name.replace('-', "_"), target.clone());
                map.insert(name.replace('_', "-"), target);
            }
        }

        // 2) system + user modules — listDir 폴백 (실패 시 silent)
        for dir in &["system/modules", "user/modules"] {
            if let Ok(entries) = self.storage.list_dir(dir).await {
                for entry in entries.iter().filter(|e| e.is_directory) {
                    let path = format!("{}/{}/index.mjs", dir, entry.name);
                    let target = CallTarget::Execute { path: path.clone() };
                    map.insert(entry.name.clone(), target.clone());
                    map.insert(entry.name.replace('-', "_"), target.clone());
                    map.insert(entry.name.replace('_', "-"), target.clone());
                    map.insert(format!("sysmod_{}", entry.name), target.clone());
                    map.insert(
                        format!("sysmod_{}", entry.name.replace('-', "_")),
                        target.clone(),
                    );
                    map.insert(path, target);
                }
            }
        }

        let result = Self::lookup_variants(&map, identifier);
        if let Ok(mut cache) = self.cache.lock() {
            *cache = Some(CachedTargets {
                map,
                cached_at: Instant::now(),
            });
        }
        result
    }

    /// 사전 승인 필요 여부 판정 — 되돌리기 어려운 작업만 user confirmation.
    /// `None` = 즉시 실행 OK. `Some(summary)` = pending action 으로 UI 표시.
    /// 옛 TS `checkNeedsApproval` 1:1.
    pub async fn check_needs_approval(&self, tc: &ToolCall) -> Option<ApprovalSummary> {
        let args = tc.arguments.as_object()?;
        match tc.name.as_str() {
            "write_file" => {
                let path = args.get("path").and_then(|v| v.as_str())?;
                // 새 파일은 즉시 작성 — 옛 TS 와 동등 (파일 존재 검사 후 수정 시만 승인)
                let exists = self.storage.read(path).await.is_ok();
                if exists {
                    Some(ApprovalSummary {
                        summary: format!("파일 수정: {}", path),
                    })
                } else {
                    None
                }
            }
            "save_page" => {
                let slug = args.get("slug").and_then(|v| v.as_str())?;
                // page builder 박혀있을 때만 검사 — 미박음 시 conservative 박지 않고 즉시 (옛 TS 동등)
                let exists = match &self.page {
                    Some(page) => page.get(slug).is_some(),
                    None => false,
                };
                if exists {
                    Some(ApprovalSummary {
                        summary: format!("페이지 수정: /{}", slug),
                    })
                } else {
                    None
                }
            }
            "delete_file" => {
                let path = args
                    .get("path")
                    .and_then(|v| v.as_str())
                    .unwrap_or("(unknown)");
                Some(ApprovalSummary {
                    summary: format!("파일 삭제: {}", path),
                })
            }
            "delete_page" => {
                let slug = args
                    .get("slug")
                    .and_then(|v| v.as_str())
                    .unwrap_or("(unknown)");
                Some(ApprovalSummary {
                    summary: format!("페이지 삭제: /{}", slug),
                })
            }
            "schedule_task" => {
                let title = args
                    .get("title")
                    .and_then(|v| v.as_str())
                    .unwrap_or("(제목 없음)");
                let when = args
                    .get("cronTime")
                    .and_then(|v| v.as_str())
                    .map(String::from)
                    .or_else(|| {
                        args.get("runAt")
                            .and_then(|v| v.as_str())
                            .map(String::from)
                    })
                    .or_else(|| {
                        args.get("delaySec")
                            .and_then(|v| v.as_i64())
                            .map(|s| format!("{}초 후", s))
                    })
                    .unwrap_or_default();
                Some(ApprovalSummary {
                    summary: format!("예약 등록: {} ({})", title, when),
                })
            }
            "cancel_task" => {
                let job_id = args
                    .get("jobId")
                    .and_then(|v| v.as_str())
                    .unwrap_or("(unknown)");
                // jobId 로 title lookup — UI 가독성 ↑ (schedule manager 박혀있을 때만)
                let label = if let Some(schedule) = &self.schedule {
                    let jobs = schedule.list();
                    let job = jobs.iter().find(|j| j.job_id == job_id);
                    match job.and_then(|j| j.options.title.as_deref()) {
                        Some(t) => format!("{} ({})", t, job_id),
                        None => job_id.to_string(),
                    }
                } else {
                    job_id.to_string()
                };
                Some(ApprovalSummary {
                    summary: format!("예약 해제: {}", label),
                })
            }
            _ => None,
        }
    }

    /// 승인 대기 도구 인자 사전 검증 — 실패 시 에러 메시지 반환 (pending 생성 전 거부).
    /// 옛 TS `preValidatePendingArgs` 1:1. 일반 로직 — pipeline step type 별 필수 필드 검증.
    pub fn pre_validate_pending_args(&self, tc: &ToolCall) -> Option<String> {
        let args = tc.arguments.as_object()?;
        match tc.name.as_str() {
            "schedule_task" => Self::validate_schedule_task(args),
            "write_file" => {
                let path = args
                    .get("path")
                    .and_then(|v| v.as_str())
                    .map(|s| s.trim())
                    .unwrap_or("");
                if path.is_empty() {
                    return Some("write_file 인자 누락: path 필수.".to_string());
                }
                if args.get("content").is_none() {
                    return Some("write_file 인자 누락: content 필수.".to_string());
                }
                None
            }
            "save_page" => {
                let slug = args
                    .get("slug")
                    .and_then(|v| v.as_str())
                    .map(|s| s.trim())
                    .unwrap_or("");
                if slug.is_empty() {
                    return Some("save_page 인자 누락: slug 필수.".to_string());
                }
                if !args.get("spec").is_some_and(|v| v.is_object()) {
                    return Some("save_page 인자 누락: spec 필수.".to_string());
                }
                None
            }
            _ => None,
        }
    }

    fn validate_schedule_task(args: &serde_json::Map<String, serde_json::Value>) -> Option<String> {
        let is_agent = args
            .get("executionMode")
            .and_then(|v| v.as_str())
            .map(|s| s == "agent")
            .unwrap_or(false);
        let has_target = args
            .get("targetPath")
            .and_then(|v| v.as_str())
            .map(|s| !s.trim().is_empty())
            .unwrap_or(false);
        let pipeline = args.get("pipeline").and_then(|v| v.as_array());
        let has_pipeline = pipeline.map(|p| !p.is_empty()).unwrap_or(false);
        let has_agent_prompt = args
            .get("agentPrompt")
            .and_then(|v| v.as_str())
            .map(|s| !s.trim().is_empty())
            .unwrap_or(false);

        if is_agent {
            if !has_agent_prompt {
                return Some(
                    "schedule_task 인자 누락: agent 모드는 agentPrompt 필수입니다. \
                     트리거 시 AI 에 전달할 자연어 instruction \
                     (잡 목적·필요 데이터·출력 형식·알림) 작성하세요."
                        .to_string(),
                );
            }
        } else if !has_target && !has_pipeline {
            return Some(
                "schedule_task 인자 누락: targetPath 또는 pipeline 중 하나는 반드시 지정해야 합니다. \
                 (agent 모드면 executionMode:\"agent\" + agentPrompt)"
                    .to_string(),
            );
        }

        let has_when = args.get("cronTime").is_some()
            || args.get("runAt").is_some()
            || args.get("delaySec").is_some();
        if !has_when {
            return Some(
                "schedule_task 인자 누락: cronTime / runAt / delaySec 중 하나는 반드시 지정해야 합니다."
                    .to_string(),
            );
        }

        // pipeline step 별 필수 필드 검증 (일반 로직 — type 별 분기)
        if let Some(pipeline) = pipeline {
            for (i, step) in pipeline.iter().enumerate() {
                let step_num = i + 1;
                let step_obj = match step.as_object() {
                    Some(o) => o,
                    None => return Some(format!("[Step {}] step이 객체가 아닙니다.", step_num)),
                };
                let step_type = step_obj.get("type").and_then(|v| v.as_str());
                let t = match step_type {
                    Some(t) if !t.is_empty() => t,
                    _ => return Some(format!(
                        "[Step {}] type 누락 — EXECUTE/MCP_CALL/NETWORK_REQUEST/LLM_TRANSFORM/CONDITION 중 하나를 지정하세요.",
                        step_num
                    )),
                };
                if !matches!(
                    t,
                    "EXECUTE" | "MCP_CALL" | "NETWORK_REQUEST" | "LLM_TRANSFORM" | "CONDITION"
                ) {
                    return Some(format!("[Step {}] 알 수 없는 type: {}", step_num, t));
                }
                if let Some(err) = Self::validate_step_fields(step_num, t, step_obj) {
                    return Some(err);
                }
            }
        }
        None
    }

    /// step type 별 필수 필드 검증 — 일반 로직 (옛 TS preValidatePendingArgs 1:1).
    fn validate_step_fields(
        step_num: usize,
        step_type: &str,
        step: &serde_json::Map<String, serde_json::Value>,
    ) -> Option<String> {
        match step_type {
            "EXECUTE" => {
                if !step.get("path").is_some_and(|v: &serde_json::Value| v.is_string()) {
                    return Some(format!(
                        "[Step {}] EXECUTE에 path 필수 (예: system/modules/kakao-talk/index.mjs).",
                        step_num
                    ));
                }
                let input_data = step.get("inputData").and_then(|v| v.as_object());
                if input_data.is_none() || input_data.unwrap().is_empty() {
                    return Some(format!(
                        "[Step {}] EXECUTE 인자 오류: 모듈 실행 파라미터는 step 평면이 아니라 \
                         inputData 객체에 넣어야 합니다. \
                         잘못: {{type:\"EXECUTE\",path:\"...\",action:\"price\",symbol:\"...\"}} · \
                         올바름: {{type:\"EXECUTE\",path:\"...\",inputData:{{action:\"price\",symbol:\"...\"}}}}",
                        step_num
                    ));
                }
                None
            }
            "MCP_CALL" => {
                if !step.get("server").is_some_and(|v: &serde_json::Value| v.is_string())
                    || !step.get("tool").is_some_and(|v: &serde_json::Value| v.is_string())
                {
                    Some(format!("[Step {}] MCP_CALL에 server, tool 필수.", step_num))
                } else {
                    None
                }
            }
            "NETWORK_REQUEST" => {
                if !step.get("url").is_some_and(|v: &serde_json::Value| v.is_string()) {
                    Some(format!("[Step {}] NETWORK_REQUEST에 url 필수.", step_num))
                } else {
                    None
                }
            }
            "LLM_TRANSFORM" => {
                if !step.get("instruction").is_some_and(|v: &serde_json::Value| v.is_string()) {
                    Some(format!(
                        "[Step {}] LLM_TRANSFORM에 instruction 필수.",
                        step_num
                    ))
                } else {
                    None
                }
            }
            "CONDITION" => {
                if !step.get("field").is_some_and(|v: &serde_json::Value| v.is_string())
                    || !step.get("op").is_some_and(|v: &serde_json::Value| v.is_string())
                {
                    Some(format!("[Step {}] CONDITION에 field, op 필수.", step_num))
                } else {
                    None
                }
            }
            _ => None,
        }
    }
}

// Tests 이관 — `infra/tests/ai_tool_dispatcher_test.rs` (integration test).
